umbral/umbral.py

```diff
@@ -120,98 +119,124 @@ def __bytes__(self):
+
+        self._point_eph_e = point_eph_e
+        self._point_eph_v = point_eph_v
+        self._bn_sig = bn_sig
+
+        self._point_eph_e_prime = e_prime
+        self._point_eph_v_prime = v_prime
+        self._point_noninteractive = noninteractive_point
 
         self.cfrags = {}
+        self._contents = None
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    class NotValid(ValueError):
+        """
+        raised if the capusle does not pass verification.
+        """
+
+    @classmethod
+    def from_bytes(cls, data: bytes, curve: ec.EllipticCurve, is_reconstructed=False):
         """
         Instantiates a Capsule object from the serialized data.
         """
-        eph_e = Point.from_bytes(data[0:33], curve)
-        eph_v = Point.from_bytes(data[33:66], curve)
-        sig = BigNum.from_bytes(data[66:98], curve)
+        if is_reconstructed:
+            e_prime = Point.from_bytes(data[0:33], curve)
+            v_prime = Point.from_bytes(data[33:66], curve)
+            eph_ni = Point.from_bytes(data[66:99], curve)
 
-        return Capsule(eph_e, eph_v, sig)
+            return cls(e_prime=e_prime, v_prime=v_prime, noninteractive_point=eph_ni)
+        else:
+            eph_e = Point.from_bytes(data[0:33], curve)
+            eph_v = Point.from_bytes(data[33:66], curve)
+            sig = BigNum.from_bytes(data[66:98], curve)
 
-    def to_bytes(self):
+            return cls(eph_e, eph_v, sig)
+
+    @property
+    def contents(self):
+        return self._contents
+
+    def to_bytes(self, reconstructed_components=False):
         """
         Serialize the Capsule into a bytestring.
         """
-        eph_e = self.point_eph_e.to_bytes()
-        eph_v = self.point_eph_v.to_bytes()
-        sig = self.bn_sig.to_bytes()
+        if reconstructed_components:
+            eph_e = self._point_eph_e_prime
+            eph_v = self._point_eph_v_prime
+            point_noninter = self._point_noninteractive
+
+            return eph_e.to_bytes() + eph_v.to_bytes() + point_noninter.to_bytes()
+        else:
+            eph_e = self._point_eph_e.to_bytes()
+            eph_v = self._point_eph_v.to_bytes()
+            sig = self._bn_sig.to_bytes()
 
-        return eph_e + eph_v + sig
+            return eph_e + eph_v + sig
 
     def verify(self, params: UmbralParameters):
 
-        e = self.point_eph_e
-        v = self.point_eph_v
-        s = self.bn_sig
+        e = self._point_eph_e
+        v = self._point_eph_v
+        s = self._bn_sig
         h = hash_to_bn([e, v], params)
 
         return s * params.g == v + (h * e)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def open(self, pub_bob, priv_bob, pub_alice, force_reopen=False, pre=None):
+        # TODO: Raise an error here if Bob has gathered enough cFrags?  See #22.
+        if self.contents and not force_reopen:
```

Comment:
I still don't see the point of the `force_reopen` logic here. What functionality is it adding or what problem is it solving?

Reply:
Yeah, we can probably nix it.

My thinking here was that, if we *don't* provide a place to do it, then users will invent a workaround using the private methods / attrs.  However, if we expressly *don't want* people to be able to do this, then yeah, let's kill it.